### ISSUE TITLE
Request: display editor pane in a tab instead of a modal window

### ISSUE THREAD

**Issue by diamondavocado** [jobisoft/HtmlSourceEditor#4]
Would be great if the HTML editor could be displayed in a tab instead of a new window.

Like ThunderHTMLEdit:
https://i.imgur.com/60gYHn5.png

**Comment by jobisoft** [comment]
I received copyrights threads from the author of ThunderHTMLEdit, so this version needs to have differences.

**Comment by diamondavocado** [comment]
> I received copyrights threads from the author of ThunderHTMLEdit, so this version needs to have differences.

Wow that is absolutely ridiculous. Like, seriously ridiculous. They have absolutely no basis for claiming copyright for several reasons:

1. Copyright is only applicable to the specific implementation of a work (i.e. the code itself), not the concept. As long as you didn't use any of ThunderHTMLEdit's *code* their threats have zero basis in reality.

2. They cannot claim to have any ownership over the concept of HTML source editing, even as it applies to Thunderbird or emails. Not in their wildest dreams. This is why software like Microsoft Word cannot have a monopoly on word processing, and why there cannot only be one type of firewall/ad-blocker/spreadsheet/calculator software.

3. They cannot claim ownership of the concept of using tabs instead of windows. The concept of tabs has been used for decades and is virtually universal among all software.

They are absolutely dreaming.

You should feel extremely comfortable in telling them to back off. If they persist in making baseless threats against you, you could report them to GitHub (or whatever) for harrassment. There are also several online communities who would back you up legally.

**Comment by Betterbird** [comment]
> I received copyrights threads from the author of ThunderHTMLEdit, so this version needs to have differences.

I'm not very happy with this statement. It was pointed out that ThunderHTMLEdit is "all rights reserved", that's all. Naturally, breaching copyright can have consequences.

TB itself already has an inbuilt HTML editor (compose window, "Insert > HTLM") so no on would claim copyright for that concept or adding tabs anywhere. As was pointed out in point 1 above, the copyright covers the specific implementation.

Technically, unless I'm mistaken, adding a tab to the compose window would require some DOM manipulations, and there is currently no WebExtension API for that. If I understand it correctly, @jobisoft clearly prefers not to venture into making his add-on a WebExtension Experiment since that would possibly require significant rework for each major TB release.

**Comment by jobisoft** [comment]
> I'm not very happy with this statement. It was pointed out that ThunderHTMLEdit is "all rights reserved", that's all. Naturally, breaching copyright can have consequences.

Do you want me to post your message here?

Edit: And you completely ignored the fact that you removed the [Mozilla Public License, version 2.0](http://www.mozilla.org/MPL/2.0/) license from the [original project](https://addons.thunderbird.net/addon/stationery/versions/1.0.0) which you forked yourself.

**Comment by Betterbird** [comment]
First of all, I'd like to focus on technical aspects here. The reason that your add-on doesn't display the HTML in a tab is that there is no API to do it. It has nothing to do with copyright issues. But since we regrettably already went beyond the technical discussion, let me say this:

IMHO, there was no need to create a new add-on since ThunderHTMLedit is available and maintained. It is unchanged from previous versions, the only change is its hosting location. You hold a different opinion, fine, but are you planning to fork [Shrunked Image Resizer](https://addons.thunderbird.net/en-GB/thunderbird/addon/shrunked-image-resizer/) as well just because it is now hosted [here](https://github.com/memeller/shrunked/releases) and the author shows no interest to host it on ATN, see [here](https://github.com/memeller/shrunked/issues/7#issuecomment-1695815651)?

The e-mail in question pointed out that parts of the code are subject to restrictive copyright and that a breach would have consequences. Calling that a threat or even (plural) threats appears exaggerated. If that were a threat, citizens worldwide would live under constant "threat" by shops which put up signs _"All shop lifting will be prosecuted"_ or in fact by their own governments or law enforcement agencies who "threaten" breaches of the law with consequences: _"If you speed and we catch you, we'll fine you"_. The exact line in my e-mail of 2nd August 2023 regrettably wasn't very well formulated, but I quote it here: _"As soon as this goes too close to ThunderHTMLedit, legal action will be taken, no doubt_".

Sadly your remark of the alleged threat has been blown out of proportion by this unfortunate [Reddit campaign](https://www.reddit.com/r/Thunderbird/comments/16nptsi/thunderbird_extension_author_and_project_leader/) almost two months later. This is how I got here. I was very surprised that two-month-old exchange has now been quoted as reason for the absence of a function which can't be implemented with the current API. And all that, because the reporter of this ticket wants his tab.

--

Re. your edit: I'm really asking myself what we are trying to achieve in an issue entitled _"Request: display editor pane in a tab instead of a modal window"_. What does the pedigree of ThunderHTMLedit have to do with the issue at hand? The original project was an add-on called Stationery that let you define templates. Its HTML source editor was experimental in the beginning and not even enabled. It later moved to using the Ace Editor, and then was abandoned at Thunderbird version 60 before the removal of support for XUL overlays. ThunderHTMLedit acknowledges its origins everywhere as correct attribution: _"Original idea by Łukasz 'Arivald' Płomiński, author of the 'Stationery' add-on"_. There is not much left of the original code, which was mostly glue between Ace and the Thunderbird composer of the XUL overlay days. Be that as it may: What are you implying with your statement? A) that you are now representing Łukasz 'Arivald' Płomiński and allege that I have breached his copyright? And B) Even that if this were the case, that would give you the right to breach copyright on the pieces which are clearly my own original work?

John, please let's get back to the issue at hand: Editing in a tab in your add-on. Nothing else.

**Comment by diamondavocado** [comment]
> IMHO, there was no need to create a new add-on since ThunderHTMLedit is available and maintained. It is unchanged from previous versions, the only change is its hosting location. 

@Betterbird Many people (myself included) prefer to exclusively use open-source software (for many different valid reasons). If you will make ThunderHTMLEdit open-source then there would be absolutely no problem. Is that something you will do?

> The exact line in my e-mail of 2nd August 2023 regrettably wasn't very well formulated, but I quote it here: "As soon as this goes too close to ThunderHTMLedit, legal action will be taken, no doubt".

In my opinion your attitude - exemplified by your quote - is what has triggered all of this.

You simply can't tell someone "your software looks too much like my software and I'm going to sue you", which is basically what you were saying. That has absolutely no legal basis, as I explained in my comment above. It's a hostile attitude which assumes bad faith and it paints you in a very poor light.

The bottom line is you have no right to threaten any kind of legal action. There is zero legal basis for legal action against someone because their software is "too close" to your software. It's a worthless statement and an empty threat.

**The bottom line is:** the only way you can bring legal action against someone for "copying" your software is:

1. If they copy and paste your code directly; AND
2. Your software was not published under a public/permissive license; AND
3. Your software was not originally derived from a project which was published under a share-alike license; AND
4. The code in question demonstrates sufficient creativity to be considered an "original work".

AND/OR

1. You are the owner of a patent which has been granted for a particularly unique approach/design/solution which your software implements.

Now, please feel free to correct me if any of the following is wrong, but my understanding is:

1. No code has been directly copied from ThunderHTMLEdit.
2. ThunderHTMLEdit was forked from a project published under the Mozilla Public License v2, which requires you to make available all portions of source code originally stemming from that project, and I am unsure if that has been done.
3. You do not own a patent for any part of ThunderHTMLEdit or the GUIs/layouts/design decisions used in the software.

**TL;DR** you can't threaten someone for making software that's "too close" to yours.

**Comment by Betterbird** [comment]
> Many people (myself included) prefer to exclusively use open-source software (for many different valid reasons). If you will make ThunderHTMLEdit open-source then there would be absolutely no problem. Is that something you will do?

ThunderHTMLEdit is open source. Open it with Zip and then an editor. You're confusing the openness of the source with the license.

As stated above, the phrasing used in a private e-mail to John was unfortunate, but the core of the statement holds true: The non-Ace portions of ThunderHTMLEdit are original work and protected by restrictive copyright. Download whatever you deem fit and compare, I sent this screenshot to John:
![image](https://github.com/jobisoft/HtmlSourceEditor/assets/86832723/43d293ce-c1b6-400d-a0e7-9964e4d869e2)

I don't wish to discuss this any further. It is totally off-topic for this GitHub issue. I don't feel obliged to make any statement on whether John has copied code or not. He told that he hadn't and I left it at that. I even pointed out how to improve his add-on further. I'm totally dismayed that discussion has come up again after two months and is being conducted in public, not to mention your Reddit campaign.

Edit: corrected typos.

**Comment by diamondavocado** [comment]
@Betterbird If you don't wish to discuss these issues, *don't make baseless threats to other developers.* Like it or not, such things have consequences.

I made a feature request. The developer is obviously hesitant to implement such a feature because of whatever interaction **you** have had with them.

> As stated above, the phrasing used in a private e-mail to John was unfortunate, but the core of the statement holds true: The non-Ace portions of ThunderHTMLEdit are original work and protected by restrictive copyright

I really want to ask: "**What is your point?**" You don't even know if the developer has even *looked* at your source code, let alone considered infringing your copyright. Until you have actual evidence of a copyright infringement, you shouldn't even be **mentioning** copyright. 

How would you like it if Microsoft emailed you and said "Hey, don't get too close to Outlook or we'll sue you into the ground"?

Bottom line is you are just pre-emptively trying to stop others from innovating because you're gatekeeping your own precious extensions. You're acting like a bully.

If you legitimately had (or have) concerns about another developer copying your work, you know what a reasonable way to approach it would be? Telling them _"Hey, great to see someone else working on a Thunderbird HTML editor... just so you know, I worked really hard on the code in my extension and I'm keeping it under copyright so please don't use any of it in your own work. Thanks."_

Instead, you've made a developer hesitant to take feature requests on their own goddamn project, and that's not cool at all.

**Comment by Betterbird** [comment]
Maybe one day you should install a Microsoft product and check their "threats", they are bullies for showing you this when you install their software:

![image](https://github.com/jobisoft/HtmlSourceEditor/assets/86832723/64e2a0d7-a33f-4b27-b230-bbf05af8c16f)

Maybe you should also not tell others what they should or shouldn't say and call them names just because they pronounced rights to their original work. Maybe also you should take a course on copyright to understand that this is about protecting ones intellectual property, which in this case doesn't even hamper innovation since it's all open source anyway.

**Comment by diamondavocado** [comment]
> Maybe one day you should install a Microsoft product and check their "threats", they are bullies for showing you this when you install their software

The obvious and massive difference is Microsoft isn't making uninvited contact with random open source developers on GitHub to pre-emptively tell them that stuff. 

Plus I've already explained in painful detail why your claim of copyright as it relates to this extension is without any legal merit.

I'm sure you've already got copyright notices in your extension's source code, right? Isn't that enough for you? Are you gonna contact every person who ever makes a Thunderbird extension and give them the same warning? C'mon dude, you're defending the indefensible.

**Comment by diamondavocado** [comment]
Anyway...

@jobisoft I understand if you can't be bothered messing around with experimental WebExtension stuff. 

Can you please keep this issue open until there exists an easier way to integrate the editor as a tab instead of a window (or until someone else submits a PR with the functionality)?

**Comment by jobisoft** [comment]
Yes, I will keep the issue open.
